core/src/test/scala/org/bykn/bosatsu/EvaluationTest.scala

```diff
@@ -1993,6 +1997,17 @@ w = ListWrapper([], True)
 
 ListWrapper([*_], r) = w
 
+tests = Assertion(r, "match with total list pattern")
+"""), "A", 1)
+
+    runBosatsuTest(List("""package A
+
+struct ListWrapper2(items: List[a], others: List[b], b: Bool)
```

Comment:
add another example with `struct ListWrapperTup(items: List[(a, b)], b: Bool)` and make sure we push down on match in there as well.